### PR TITLE
rpc: Use the blocks pinning API for chainHead methods

### PR DESCRIPTION
**Pull request by lexnv**
This PR uses the new pinning API introduced by https://github.com/paritytech/substrate/pull/13157.

The backend is propagated to the subscription management, which internally calls pinning only once for a given block.
An alternative to this approach would be holding the `UnpinHandle` structures that perform the unpinning on drop.
I decided to implement the first option because`chainHead` must also report all the blocks from memory and those blocks are not generated by notifications, therefore having a single uniform way of pinning across the chainHead methods.

To ensure proper testing the in-memory `Backend` is extended to describe the number of active pinned references for a given block. An in-memory `Backend` instance is used for testing the `chainHead`, therefore the `children` method of the in-memory Backend is extended to return an empty list.

### CODE REVIEW

File: client/rpc-spec-v2/src/chain_head/chain_head.rs [hunk 1]
```diff
@@ -60,15 +60,15 @@ use sp_runtime::{
 use std::{marker::PhantomData, sync::Arc};
 
 /// An API for chain head RPC calls.
-pub struct ChainHead<BE, Block: BlockT, Client> {
+pub struct ChainHead<BE: Backend<Block> + 'static, Block: BlockT, Client> {
```
[the remark this answers is not in the document]
Comment (lexnv) [reply client/rpc-spec-v2/src/chain_head/chain_head.rs:63]: Yep, I do like that approach. One nice thing about not having bounds on the structure is that we do not contaminate every structure containing that type.

For soundness reasons I presume, the compiler requires the bounds for the `Drop` impl to also be present on the struct. 
The `SubscriptionHandle` implements `Drop` that needs access to the `unpin_block` method defined by `Backend<Block>` trait.
And unfortunately, this then leaks into `SubscriptionManagement` and into `ChainHead` itself.

[the remark this answers is not in the document]
Comment (jsdw) [reply client/rpc-spec-v2/src/chain_head/chain_head.rs:63]: Thanks for this ; TIL `Drop` impls with trait bounds need matching ones on the struct :)

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 2]
```diff
@@ -132,14 +153,26 @@ impl<Block: BlockT> SubscriptionHandle<Block> {
...
 
+impl<Block: BlockT, BE: Backend<Block> + 'static> Drop for SubscriptionHandle<Block, BE> {
+	fn drop(&mut self) {
+		let mut inner = self.inner.write();
+
+		// Unpin any remaining blocks of this subscription.
+		for hash in &inner.blocks {
+			self.backend.unpin_block(*hash);
+		}
+		inner.blocks.clear();
```
Comment (bkchr) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:164]:
```suggestion
		for hash in inner.blocks.drain() {
			self.backend.unpin_block(*hash);
		}
```

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 3]
```diff
@@ -94,13 +104,17 @@ impl<Block: BlockT> SubscriptionHandle<Block> {
 
 		if inner.blocks.len() == inner.max_pinned_blocks {
 			// We have reached the limit. However, the block can be already inserted.
+
```
Comment (bkchr) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:107]:
```suggestion
			// We reached the limit. However, the block could already be pinned.

```

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 4]
```diff
@@ -94,13 +104,17 @@ impl<Block: BlockT> SubscriptionHandle<Block> {
...
 			if inner.blocks.contains(&hash) {
```
Comment (bkchr) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:108]: Here is a logic error? If we try to pin the same block twice, the first `unpin` will unpin it in the backend (assuming it is only pinned by this subscription). Is this documented in the spec that when you receive the same block as best and some seconds later as finalized (without having called unpin in between), you only need to call unpin once?

**Review by bkchr — COMMENTED** [review]
Looks good in general, just some small things that need to be improved!

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2325858

**Comment by skunert** [comment]
The block limit is currently set to 4096 here: https://github.com/paritytech/substrate/blob/901506114ec3102d245049ab0f89d89190e30c21/client/service/src/builder.rs#L703
The limit for the pinned block cache in the backend however is 1024 (arbitrarily chosen):
https://github.com/paritytech/substrate/blob/901506114ec3102d245049ab0f89d89190e30c21/client/db/src/pinned_blocks_cache.rs#L23
We should adjust these to they match, preferably keeping  the backend one larger than the RPC limit, since it needs to keep additional headroom for notification pinning.

Also I would like to discuss the impact on node memory:
- RPC API has a limit of pinned blocks per subscription
- BlockchainDB has a limit of overall pinned blocks for justification & body
- Afaik State-DB does currently not have a hard limit of pinned blocks
- So theoretically multiple clients could try to keep their maximum of allowed block pins in non-overlapping windows which would increase the nodes memory usage. At some point, BlockchainDB would stop pinning, but StateDB would not. Also when the limit is reached in BlockchainDB, RPC layer would think that a block is pinned when it actually is not. We should probably implement some feedback whether pinning in the backend was sucessful.
- Maybe we should also globally keep track of pinned blocks in the RPC API
- We could also implement some generous limit in the State-DB for the pinning.

What do you guys think?

File: client/rpc-spec-v2/src/chain_head/subscription.rs (hunk 4 above)
Comment (lexnv) [reply client/rpc-spec-v2/src/chain_head/subscription.rs:108]: Nice catch! I couldn't find something super specific in the RPC spec regarding this.

From the [spec](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html):
> Note: A JSON-RPC client should call chainHead_unstable_unpin only after it is sure to no longer be interested in a certain block. This typically happens after the block has been finalized or pruned. There is no requirement to call chainHead_unstable_unpin as quickly as possible.

I was also thinking along the lines: the [newBlock](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#newblock) event is where the block is declared and pinned; and [bestBlockChanged](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#bestblockchanged) and [finalized](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#finalized) are updates about what happens with that block.

I think we could update the spec documentation for the `unpin` method to state that only block hashes reported by the `newBlock` event are supposed to be unpinned.

What do you think about it?

Comment (jsdw) [reply client/rpc-spec-v2/src/chain_head/subscription.rs:108]: > I was also thinking along the lines: the [newBlock](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#newblock) event is where the block is declared and pinned; and [bestBlockChanged](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#bestblockchanged) and [finalized](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#finalized) are updates about what happens with that block.

It sounds reasonable to me that:
- a block is pinned just prior to a client being given a `newBlock` event, 
- later we might see `best`/`finalised` events emitted re that block (and no pinning happens at this point), 
- at any point after the `newBlock`, whenever a user would like to unpin that block hash they will call "unpin" once against the hash to unpin it.

But, if a `newBlock` is unpinned quickly and _then_ a best/finalized event is emitted for it.. should it need unpinning again? I think offhand I'd go with "it's been unpinned so it stays unpinned, and those events are just notifications about the block you've already declared you don't care about any more".

But that would make it annoying for people to ignore all but best/finalized blocks! Because the simplest logic to do so would be to call "unpin" on all "newBlock" events and not on any best/finalized block events.

"unpin" could potentially take both a block hash _and_ a type (new/best/finalized) so that you can opt-out of all block events you don't care about but still keep pinned the ones you do?

Or, a best/finalized notificiation could pin the block again, and "unpin" just takes the hash, but that runs the risk of race conditions (I "unpin" a newBlock but by the time the unpin request hits the node ,a bestBlock for that hash is emitted).

So two options then are:
- Blocks are _only_ pinned at "newBlock". Unpin will unpin, and best/finalized events may exist for blocks that have already been unpinned.
- new/best/finalized blocks are treated as independent. "unpin" takes a hash and type to unpin exactly one of those.

Comment (bkchr) [reply client/rpc-spec-v2/src/chain_head/subscription.rs:108]: > I was also thinking along the lines: the [newBlock](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#newblock) event is where the block is declared and pinned; and [bestBlockChanged](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#bestblockchanged) and [finalized](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_unstable_follow.html#finalized) are updates about what happens with that block.

I think we should pin on the first event we see (a client can only see bestBlock/finalized without newblock by having connected after the block was imported) and then unpin should be called when they are not interested anymore. We should update the spec to mention this behavior.

Comment (lexnv) [reply client/rpc-spec-v2/src/chain_head/subscription.rs:108]: Yep, we could update the spec to reflect something along the lines:
"Any block hash reported by chainHead's events is guaranteed to be pinned in memory and it requires a single `chainHead_unstable_unpin` call to unpin the block."

The spec states currently:
> The current finalized block reported in the initialized event, and each subsequent block reported with a newBlock event, is automatically considered by the JSON-RPC server as pinned. 

---
> I think we should pin on the first event we see

Yep, that's along the lines of how we pin internally for this `chainhead` subscription.

> a client can only see bestBlock/finalized without newblock by having connected after the block was imported

Before using the subscriptions (best block and finalized) we are:
- generating finalized event (and pin the finalized block)
- generate `newBlock` event for each children of the finalized recursively (and pin the block)

I might be wrong, but I think its entirely possible that we can miss blocks (warp sync etc) and such we always `pin` the block on the first event that we get from the subscriptions. 
In case of errors or jumps, the chainHead should produce the `Stop` events. 

Let me know if that is correct, I ll take an action to create a PR to clarify a bit the spec, thanks!

**Comment by lexnv** [comment]
The chainhead limit (4k blocks) is also arbitrarily chosen.

From the initial "pinning in database" from https://github.com/paritytech/substrate/pull/12497, we internally picked a limit of delaying the pruning by 32 blocks; considering the production of blocks at 6 seconds that would give users roughly 3 minutes to execute their use-case using the `chainHead` before considering the `archive` methods.

I like the idea of keeping in sync these values, maybe even a bit more restrictive for the chainHead.
For a more informed decision about the hard limit (pinning cache in the backend), I would try to rerun the memory profiling from https://github.com/paritytech/substrate/pull/12497 and see how much memory that would eat up (or come up with a sensitive approximation for "justification & body" size). 

To allow fairness for multiple subscriptions, in the RPC layer:
- add a hard limit of "1024" pinned blocks, tracking which subscription is active for those blocks
- add a soft limit of "64" blocks (approx 6 minutes) 
  - inform the `chainHead` spec about this
  - inform the `chainHead` that a user is encouraged to call `unpin` only once when they receive the `finalized` event (extracted from the above comment)

When a user first connects, the `chainHead` checks that we have at least "64" free blocks. 
If the hard limit is hit for pinning, drop a subscription that exceeded the soft limit and check again.
Ideally, all users will keep a maximum of 64 latest blocks pinned; and that would reflect in approx 64 blocks used out of the hard limit.
As a next step, we could try to punish subscriptions that are keeping older blocks around: when we need space we'd drop a subscription that exceeds the soft limit and has the oldest blocks (ie have a scoring value based on blocks pinned: subscription_score = Sum(each block number):  we d drop the subscription with the lowest value first).

 
> RPC layer would think that a block is pinned when it actually is not

Ideally, we'd need to submit the `event::Stop` on the subscription when we can't hold the pinning guarantees. I also think that we could send the `stop` "lazily" when the user tries to do something with that block and we realize that's nowhere to be found (we do this currently for headers because we had no true pinning IIRC).

I'm not quite sure what's the best solution to implement considering the alternatives. @bkchr what do you think?

File: client/rpc-spec-v2/src/chain_head/subscription.rs (hunk 4 above)
Comment (lexnv) [reply client/rpc-spec-v2/src/chain_head/subscription.rs:108]: Another way would be to allow users to call `unpin` only for hashes reported by the `Finalized` event: https://github.com/paritytech/json-rpc-interface-spec/issues/32. Let me know if this makes a bit more sense, thanks!

Comment (jsdw) [reply client/rpc-spec-v2/src/chain_head/subscription.rs:108]: > and then unpin should be called when they are not interested anymore.

What if they call `unpin` after receiving eg a `newBlock`, and then the server sends a `finalized` with the same hash? I assume that the block is still unpinned now.

If I was only interested in finalized blocks, and did not want to have any non-finalized blocks pinned any longer than necessary, what would I do?

I gather from chatting with Alex that that what I'd do is hold off on calling "unpin" on anything until I see a `finalized` notification (which also lists all blocks that are "pruned"), and then I could use that list to `unpin` all blocks that are listed as "pruned" now that I know they aren't important any more. (This relies on that list of pruned blocks really just being a list of all blocks not in the finalized chain, since if they are pinned they wouldn't be pruned). Does that make sense?

**Comment by skunert** [comment]
This makes a lot of sense to me. Sounds like a good compromise between resource usage and user convenience. As long as users stay within their 64 (or whatever limit we choose), they are pretty safe. Outside of this, there is a chance of getting thrown out.

> inform the chainHead that a user is encouraged to call unpin only once when they receive the finalized event (extracted from the above comment)

I am not sure what you mean by this? Why does it make a difference for us when a user calls unpin (as long as limits are met)?

> (we do this currently for headers because we had no true pinning IIRC)

There is no pinning for headers because we never prune them, I think we keep them around forever.

**Comment by lexnv** [comment]
> I am not sure what you mean by this? Why does it make a difference for us when a user calls unpin (as long as limits are met)?

Yep, indeed this doesn't make a difference. For context, I was referring to the https://github.com/paritytech/json-rpc-interface-spec/issues/32 possible race, and since we could handle that inside substrate it shouldn't matter in the end :)

**Comment by lexnv** [comment]
bot rebase

File: client/rpc-spec-v2/src/chain_head/chain_head.rs [hunk 5]
```diff
@@ -203,18 +204,26 @@ where
 		let subscriptions = self.subscriptions.clone();
 
 		let fut = async move {
-			let Some(handle) = subscriptions.get_subscription(&follow_subscription) else {
-				// Invalid invalid subscription ID.
-				let _ = sink.send(&ChainHeadEvent::<String>::Disjoint);
-				return
+			let _block_guard = match subscriptions.lock_block(&follow_subscription, hash) {
```
Comment (skunert) [inline client/rpc-spec-v2/src/chain_head/chain_head.rs:207]: Just for clarification: We only guard here to complete the current operation in case the subscription unpins in the meantime (assuming its the last subscription pinning)?

**Review by skunert — COMMENTED** [review]
Overall the logic looks good to me.  Just silly nits. I found that the bool usage made it a bit hard to follow the logic, even if nicely documented.

File: client/rpc-spec-v2/src/chain_head/chain_head.rs (hunk 5 above)
Comment (lexnv) [reply client/rpc-spec-v2/src/chain_head/chain_head.rs:207]: Yep, It may be possible, although highly unlikely that the last subscription of the block calls unpin while we perform some operation with that block.

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 6]
```diff
@@ -20,266 +20,1045 @@
...
+///
+/// ```ignore
+///                   (register)
+/// [ REGISTERED ]  ---------------> [ FULLY REGISTERED ]
+///       |                              |
+///       | (unpin)                      | (unpin)
+///       |                              |
+///       V           (register)         V
+/// [ UNPINNED ]  -----------------> [ FULLY UNPINNED ]
+/// ```
```
Comment (skunert) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:103]: Nice!

**Review by skunert — APPROVED** [review]

File: client/api/src/in_mem.rs [hunk 7]
```diff
@@ -787,11 +803,17 @@ where
 		false
 	}
 
-	fn pin_block(&self, _: <Block as BlockT>::Hash) -> blockchain::Result<()> {
+	fn pin_block(&self, hash: <Block as BlockT>::Hash) -> blockchain::Result<()> {
+		let mut blocks = self.pinned_blocks.write();
+		blocks.entry(hash).and_modify(|counter| *counter += 1).or_insert(1);
```
Comment (jsdw) [inline client/api/src/in_mem.rs:808]: Nit: I think you could write:

```rust
*blocks.entry(hash).or_default() += 1;
```

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 8]
```diff
@@ -20,266 +20,1045 @@
...
+#[derive(Debug, Clone, PartialEq)]
+enum BlockStateMachine {
+	/// The block was registered by one event (either `Finalized` or `BestBlock` event).
```
Comment (jsdw) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:106]: Just to check my understanding, I thought only `BestBlock` would lead to the `Registered` state, and then `Finalized` leads to `FullyRegistered`  (assuming unpin hasn't been called in the meantime)?

So the names could also be, in my head:

```rust
BestBlock,
FinalizedBlock,
UnpinnedBestBlock,
UnpinnedFinalizedBlock
```

Or am I wrong in my understanding?

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 9]
```diff
@@ -20,266 +20,1045 @@
...
+	/// Track the block hashes available for this subscription.
+	///
+	/// This implementation assumes:
+	/// - most of the time subscriptions keep a few blocks of the chain's head pinned
+	/// - iteration through the blocks happens only when the hard limit is exceeded.
+	///
+	/// Considering the assumption, iterating (in the unlike case) the hashmap O(N) is
+	/// more time efficient and code friendly than paying for:
+	/// - extra space: an extra BTreeMap<Instant, Hash> to older hashes by oldest insertion
+	/// - extra time: O(log(N)) for insert/remove/find each `pin` block time per subscriptions
```
Comment (jsdw) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:176]: I guess we are relying solely on users to unpin blocks properly though if we want to uphold these assumptions?

Or will I, as a user, hit the "stop" condition of "too many pinned blocks", which would lead to the subscription closing, before this needs to worry about the hard limit being hit and iteration needed?

(basically I'm wondering if there's a case where users can cause lots of iteration over this if they forget somehting or are in some way malicious?)

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 10]
```diff
@@ -20,266 +20,1045 @@
...
+	///
+	/// Returns:
+	/// - true if this is the first time that the block is registered
+	/// - false if the block was already registered
+	fn register_block(&mut self, hash: Block::Hash) -> bool {
+		match self.blocks.entry(hash) {
+			Entry::Occupied(mut occupied) => {
+				let block_state = occupied.get_mut();
+
+				block_state.state_machine.advance_register();
```
Comment (jsdw) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:203]: Nit: the name isn't the clearest to me here, but then I wonder in general whether we want to uphold the flow that a block is first registered as "best" and then next as "finalized", and in this case I wonder whether a general `register_block` function is a little vague and it might be better to have:

```rust
fn register_best_block(..)
fn register_finalized_block(..)
```

Just to enforce this ordering a little better (functions can deny if not called in the right order or whatever).

But yeah just a thought :)

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 11]
```diff
@@ -20,266 +20,1045 @@
...
+				true
+			},
+		}
+	}
+
+	/// A block is unregistered when the user calls `unpin`.
+	///
+	/// Returns:
+	/// - true if the block can be unpinned.
+	/// - false if the subscription does not contain the block.
```
Comment (jsdw) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:228]: Or if the block was already unpinned by the looks of it :)

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 12]
```diff
@@ -20,266 +20,1045 @@
...
+		let Some(state) = self.blocks.get(&hash) else {
+			// Block was not tracked.
+			return false
+		};
+
+		// Subscription no longer contains the block if `unpin` was called.
+		!state.state_machine.was_unpinned()
+	}
+
+	/// Get the timestamp of the oldest inserted block.
```
Comment (jsdw) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:266]: Worth a warning that this will iterate all of the blocks to do so?

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 13]
```diff
@@ -20,266 +20,1045 @@
...
+		Ok(())
 	}
 
-	/// Check if the block hash is present for the provided subscription ID.
-	///
-	/// Returns `true` if the set contains the block.
-	pub fn contains_block(&self, hash: &Block::Hash) -> bool {
-		let inner = self.inner.read();
-		inner.blocks.contains(hash)
+	fn global_unpin_block(&mut self, hash: Block::Hash) {
```
Comment (jsdw) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:488]: Something I find a little hard to follow is that  none of these fns are `pub` etc for a bunch of these things, and so it's hard to know which are used by the "parent" interface and which are private and only called from other fns on that type.

I wonder whether moving some of this stuff into separate files would help make the inner and outer interfaces super obvious so that it's easier to see what can and can't be called by the other types?

File: client/rpc-spec-v2/src/chain_head/subscription.rs (hunk 10 above)
Comment (jsdw) [reply client/rpc-spec-v2/src/chain_head/subscription.rs:203]: Ah I see now I've looked down; this is all exposed via `pin_block` and `unpin_block` essentially, and so yeah, I guess the idea is simply that `pin_block` can be called as many times as wanted (but at least twice; once for best and once for Finalized) and then`unpin_block` will ensure the block isn't re-pinned once it's been called (and will actually also remove the block entirely if either of these happens:

- `pin_block()`, `pin_block()`, `unpin_block()`
- `pin_block()`, `unpin_block()`, `pin_block()`

It's a little hard to "kno" what will happen I feel based on the calls, but I get that ultimately there is one pin and one unpin method that we expose, and so things have to just be correct behind that interface re best and finalized blocks

**Review by jsdw — APPROVED** [review]
Looks great! Just a couple of small comments/suggestions for clarity but nothing super important (I resolved a couple of my own comments as I went through and understood it better)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2719408

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 14]
```diff
@@ -20,266 +20,1049 @@
...
-			} else {
+		// Terminate all subscriptions that have blocks older than
+		// the specified pin duration.
+		let now = Instant::now();
+
+		let to_remove: Vec<_> = self
+			.subs
+			.iter_mut()
+			.filter_map(|(sub_id, sub)| {
+				let sub_time = sub.find_oldest_block_timestamp();
```
Comment (davxy) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:414]: If we have to choose which subscription to remove... shouldn't we have to find the newest block timestamp for a subscription? What I mean is if I have several subscriptions is not the most recent one that should be considered? Otherwise I risk to lower my pinning time for a new request just because an old one I previously made

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 15]
```diff
@@ -20,266 +20,1049 @@
...
+		// Ensure we have enough space only if the hash is not globally registered.
+		if !self.global_blocks.contains_key(&hash) {
+			// Subscription ID was terminated while ensuring enough space.
+			if self.ensure_block_space(sub_id) {
 				return Err(SubscriptionManagementError::ExceededLimits)
 			}
 		}
 
-		Ok(inner.blocks.insert(hash))
+		self.global_pin_block(hash)?;
```
Comment (davxy) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:472]: maybe this `global_pin_block` action can be moved within the `!self.global_blocks.contains_key(&hash)` scope

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 16]
```diff
@@ -20,266 +20,1049 @@
...
+// Blockchain error does not implement `PartialEq` needed for testing.
+impl PartialEq for SubscriptionManagementError {
+	fn eq(&self, other: &SubscriptionManagementError) -> bool {
+		match (self, other) {
+			(Self::ExceededLimits, Self::ExceededLimits) |
+			// Not needed for testing.
+			(Self::Blockchain(_), Self::Blockchain(_)) |
+			(Self::BlockHashAbsent, Self::BlockHashAbsent) |
+			(Self::SubscriptionAbsent, Self::SubscriptionAbsent) => true,
+			(Self::Custom(lhs), Self::Custom(rhs)) => lhs == rhs,
```
Comment (davxy) [inline client/rpc-spec-v2/src/chain_head/subscription.rs:66]: Missing `BlockHeaderAbsent` case

File: client/rpc-spec-v2/src/chain_head/subscription.rs [hunk 17]
```diff
@@ -20,266 +20,1049 @@
...
+	) -> Result<BlockGuard<Block, BE>, SubscriptionManagementError> {
+		let Some(sub) = self.subs.get(sub_id) else {
+			return Err(SubscriptionManagementError::SubscriptionAbsent)
+		};
+
+		if !sub.contains_block(hash) {
+			return Err(SubscriptionManagementError::BlockHashAbsent)
+		}
+
+		BlockGuard::new(hash, sub.runtime_updates, self.backend.clone())
```
[the remark this answers is not in the document]
Comment (lexnv) [reply client/rpc-spec-v2/src/chain_head/subscription.rs:537]: Yep, the `lock_block` function bypasses the chainHead pinning counter.

The `pin_block` reference counts the block once per subscription and will increase the `counter` only on the first call. The other calls are advancing the state machine of the block.

We probably could use `global_pin_block` and `global_unpin_block`, however, we'd need a reference to the `SubscriptionManagement` or add an extra lock and pass in a reference to `SubscriptionsInner`.

That felt more involved and I've relied on the backend's pinning reference count for this

File: client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 18]
```diff
@@ -310,7 +310,7 @@ where
 		startup_point: &StartupPoint<Block>,
 	) -> Result<Vec<FollowEvent<Block::Hash>>, SubscriptionManagementError> {
 		// The block was already pinned by the initial block events or by the finalized event.
-		if !self.sub_handle.pin_block(notification.hash)? {
+		if !self.sub_handle.pin_block(&self.sub_id, notification.hash)? {
```
[the remark this answers is not in the document]
Comment (lexnv) [reply client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:313]: Yep, we keep track internally of the calls made to the `pin_block`.
The second call will return `false` to signal we generated the events for this block. 

The `pin_block` can happen from the
- import blocks 
- finalized blocks

We use the `BlockStateMachine` to ensure that we untrack a block from memory only when we receive two calls to the `pin_block` and one `unpin_block` . 

This will protect us against:
- import event comes in: we generate events
- user calls unpin: we previously lost track of the block
- finalized event comes in: we previously regenerated events

However if both events come in (import and finalized) and the user calls unpin; _and_ the backend gives us the same block again hash again (most probably a bug), we do not protect against this :D

File: client/rpc-spec-v2/src/chain_head/subscription.rs (hunk 15 above)
Comment (davxy) [reply client/rpc-spec-v2/src/chain_head/subscription.rs]: No I  was thinking about :
```
if !self.global_blocks.contains_key(&hash) {
     if self.ensure_block_space(sub_id) {
         return Err...
     }
     self.global_pin_block(hash)?;
}
```

Comment (lexnv) [reply client/rpc-spec-v2/src/chain_head/subscription.rs]: Ahh I see, the `global_pin_block` function name is a bit misleading, under the hood it does backend pinning if the block does not exist or increments the internal counter if the block exists. 
This is similar to the `global_unpin_block`.

Will renames those to `global_register_block` and `global_unregister_block` similar to `SubscriptionState` naming, and add a couple of documentation :D

**Review by davxy — APPROVED** [review]
